### PR TITLE
new resource `azurerm_elastic_san_volume`

### PR DESCRIPTION
**Pull request by teowa**
reference:
1. [Azure Doc](https://learn.microsoft.com/en-us/azure/storage/elastic-san/elastic-san-create?tabs=azure-portal#create-volumes)
2. [Azure REST API specs](https://github.com/Azure/azure-rest-api-specs/blob/5e060c76edd4fab38e9202055062154006463018/specification/elasticsan/resource-manager/Microsoft.ElasticSan/stable/2023-01-01/elasticsan.json#L741)

### CODE REVIEW

File: internal/services/elasticsan/elastic_san_volume_resource.go [hunk 1]
```diff
@@ -0,0 +1,319 @@
...
+		},
+	}
+}
+
+func (r ElasticSANVolumeResource) CustomizeDiff() sdk.ResourceFunc {
+	return sdk.ResourceFunc{
+		Timeout: 5 * time.Minute,
+		Func: func(ctx context.Context, metadata sdk.ResourceMetaData) error {
+			if oldVal, newVal := metadata.ResourceDiff.GetChange("size_in_gib"); newVal.(int) < oldVal.(int) {
+				return fmt.Errorf("new size_in_gib should be greater than the existing one")
```
Comment (ms-zhenhua) [inline internal/services/elasticsan/elastic_san_volume_resource.go:143]: shall we add a testcase for this validation?

File: website/docs/r/elastic_san_volume.html.markdown [hunk 2]
```diff
@@ -0,0 +1,150 @@
...
+
+## Arguments Reference
+
+The following arguments are supported:
+
+* `name` - (Required) Specifies the name of this Elastic SAN Volume. Changing this forces a new resource to be created.
+
+* `volume_group_id` - (Required) Specifies the Volume Group ID within which this Elastic SAN Volume should exist. Changing this forces a new resource to be created.
+
+* `size_in_gib` - (Required) Specifies the size of the Elastic SAN Volume in GiB. The size should be within the remaining capacity of the parent Elastic SAN. Possible value is integer between `1` and `65536` (16 TiB).
```
Comment (ms-zhenhua) [inline website/docs/r/elastic_san_volume.html.markdown:105]:
```suggestion
* `size_in_gib` - (Required) Specifies the size of the Elastic SAN Volume in GiB. The size should be within the remaining capacity of the parent Elastic SAN. Possible values are between `1` and `65536` (16 TiB).
```

**Review by ms-zhenhua — CHANGES_REQUESTED** [review]
Hi @teowa, 
Thanks for this PR - I've taken a look through and left some comments inline. If we can fix those up, this should be good to go 👍
BTW, please also involve the test results.

File: internal/services/elasticsan/elastic_san_volume_resource.go [hunk 3]
```diff
@@ -0,0 +1,319 @@
...
+						Required: true,
+						ValidateFunc: validation.StringInSlice([]string{
+							// None is not exposed
+							string(volumes.VolumeCreateOptionDisk),
+							string(volumes.VolumeCreateOptionDiskRestorePoint),
+							string(volumes.VolumeCreateOptionDiskSnapshot),
+							string(volumes.VolumeCreateOptionVolumeSnapshot),
+						},
+							false),
+					},
```
Comment (ms-zhenhua) [inline internal/services/elasticsan/elastic_san_volume_resource.go:107]:
```suggestion
					"source_id": {
						Type:     pluginsdk.TypeString,
						Required: true,
						ForceNew: true,
						ValidateFunc: validation.Any(
							commonids.ValidateManagedDiskID,
							restorepoints.ValidateRestorePointID,
							diskSnapshots.ValidateSnapshotID,
							snapshots.ValidateSnapshotID,
						),
					},
					"source_type": {
						Type:     pluginsdk.TypeString,
						Required: true,
						ForceNew: true,
						ValidateFunc: validation.StringInSlice([]string{
							// None is not exposed
							string(volumes.VolumeCreateOptionDisk),
							string(volumes.VolumeCreateOptionDiskRestorePoint),
							string(volumes.VolumeCreateOptionDiskSnapshot),
							string(volumes.VolumeCreateOptionVolumeSnapshot),
						},
							false),
					},
```

**Comment by teowa** [comment]
image

File: internal/services/elasticsan/elastic_san_volume_resource.go (hunk 1 above)
Comment (teowa) [reply internal/services/elasticsan/elastic_san_volume_resource.go:141]: This is included in update test

**Review by ms-zhenhua — APPROVED** [review]

**Review by katbyte — CHANGES_REQUESTED** [review]
Have a test failure:
```
------- Stdout: -------
=== RUN   TestAccElasticSANVolume_complete
=== PAUSE TestAccElasticSANVolume_complete
=== CONT  TestAccElasticSANVolume_complete
    testcase.go:113: Step 1/2 error: Error running apply: exit status 1
        
        Error: creating Volume (Subscription: "*******"
        Resource Group Name: "acctestrg-esvg-240229223355576575"
        Elastic San Name: "acctestes-4pwq0"
        Volume Group Name: "acctestesvg-4pwq0"
        Volume Name: "acctestesv-4pwq0"): polling after Create: polling failed: the Azure API returned the following error:
        
        Status: "InternalServerError"
        Code: ""
        Message: "Internal server error."
        Activity Id: ""
        
        ---
        
        API Response:
        
        ----[start]----
        {"status":"Failed","error":{"code":"InternalServerError","message":"Internal server error."}}
        -----[end]-----
        
        
          with azurerm_elastic_san_volume.test,
          on terraform_plugin_test.tf line 70, in resource "azurerm_elastic_san_volume" "test":
          70: resource "azurerm_elastic_san_volume" "test" {
        
        creating Volume (Subscription: "*******"
        Resource Group Name: "acctestrg-esvg-240229223355576575"
        Elastic San Name: "acctestes-4pwq0"
        Volume Group Name: "acctestesvg-4pwq0"
        Volume Name: "acctestesv-4pwq0"): polling after Create: polling failed: the
        Azure API returned the following error:
        
        Status: "InternalServerError"
        Code: ""
        Message: "Internal server error."
        Activity Id: ""
        
        ---
        
        API Response:
        
        ----[start]----
        {"status":"Failed","error":{"code":"InternalServerError","message":"Internal
        server error."}}
        -----[end]-----
        
--- FAIL: TestAccElasticSANVolume_complete (184.96s)
FAIL
```

**Comment by teowa** [comment]
Hi, the service team should have the backend fixed, the the test can pass now:
image

**Review by katbyte — APPROVED** [review]
LGTM 🦺

**Pull request merged**